source/versioned-api/versioned-api.rst

```diff
@@ -0,0 +1,273 @@
+===================
+Unified Test Format
+===================
+
+:Spec Title: Versioned API
+:Spec Version: 1.0.0
+:Author: Andreas Braun
+:Advisors: Jeff Yemin, A. Jesse Jiryu Davis, Patrick Freed, Oleg Pudeyev
+:Status: Draft
+:Type: Standards
+:Minimum Server Version: N/A
+:Last Modified: 2020-11-05
+
+.. contents::
+
+--------
+
+Abstract
+========
+
+As MongoDB moves toward more frequent releases (a.k.a. continuous delivery), we
+want to enable users to take advantage of our rapidly released features, without
+exposing applications to incompatible server changes due to automatic server
+upgrades. A versioned API will help accomplish that goal.
+
+
+META
+====
+
+The keywords "MUST", "MUST NOT", "REQUIRED", "SHALL", "SHALL NOT", "SHOULD",
+"SHOULD NOT", "RECOMMENDED", "MAY", and "OPTIONAL" in this document are to be
+interpreted as described in `RFC 2119 <https://www.ietf.org/rfc/rfc2119.txt>`__.
+
+This document tends to use "SHOULD" more frequently than other specifications,
+but mainly in the context of providing guidance on writing test files. This is
+discussed in more detail in `Design Rationale`_.
+
+
+Specification
+=============
+
+Background
+----------
+
+When applications interact with MongoDB, both the driver and the server
+participate in executing operations. Therefore, when determining application
+compatibility with MongoDB, both the driver and the server behavior must be
+taken into account.
+
+An application can specify the server API version when creating MongoClient.
+When this is done:
+
+- The client sends the specified API version to the server, causing the server
+  to behave in a manner compatible with that API version.
+- The driver will behave in a manner compatible with the driver's behavior as of
+  the release when the driver first started to support the specified server API
+  version.
+
+Presently there is no specification for how a driver must behave when a
+particular server API version is requested, or what driver operations are
+subject to API compatibility guarantees. Such requirements may be stipulated in
+subsequent specifications.
+
+This specification requires MongoClient to validate that it supports the
+specified server API version, if any, but does not define what such support
+means.
+
+
+``MongoClient`` changes
+-----------------------
+
+``MongoClient`` instances accept a new ``serverApi`` option to allow the user to
+declare an API version:
+
+.. code:: typescript
+
+   class MongoClient {
+       MongoClient(... serverApi: ServerApi);
+   }
+
+   enum ServerApiVersion {
+       v1 = "1",
+   }
+
+   class ServerApi {
+       version: string|ServerApiVersion;
+       strict: Optional<Boolean>; // Default false
+       deprecationErrors: Optional<Boolean>; // Default false
+   }
+
+Drivers SHOULD group the ``serverApi`` option with other similar client options
+like ``autoEncryptionOpts``.
```

Comment:
can we add an explicit "Drivers MUST NOT allow specification of any versioned API options via the connection string" (or something like that) here? I think the first time this requirement shows up in this document is in the rationale, so we should probably call it out more directly here.

Reply:
Yep, seems sensible to explicitly mention this here 👍